File: python/RappCloud/CloudMsgs/EmailSend.py
from RappCloud.Objects import (
    File,
    Payload)

from Cloud import (
    CloudMsg,
    CloudRequest,
    CloudResponse)


class EmailSend(CloudMsg):
    """ Email Send Cloud Message object """

    class Request(CloudRequest):
        """ Email Send Cloud Request object.

        EmailSend.Request
        """
        def __init__(self, **kwargs):
            """!
            Constructor

            @param **kwargs - Keyword arguments. Apply values to the request attributes.
                - @ref email
                - @ref password
                - @ref server
                - @ref port
                - @ref recipients
                - @ref body
                - @ref subject
                - @ref attach_file
            """

            ## The user's email username
            self.email = ''
            ## The user's email password
            self.password = ''
            ## The email server's smtp address, i.e. 'smtp.gmail.com'
            self.server = ''
            ## The email server imap port. Leave empty to use default value.
            self.port = ''
            ## Email addresses of the recipients
            self.recipients = []
            ## The email body
            self.body = ''
            ## The email subject
            self.subject = ''
            ## Attachment file path. Can be a .zip file that will be decompressed on the server.
            self.attach_file = ''

            super(EmailSend.Request, self).__init__(**kwargs)


        def make_payload(self):
            """ Create and return the Payload of the Request. """
            return Payload(
                email=self.email,
                passwd=self.password,
                server=self.server,
                port=self.port,
                recipients=self.recipients,
                body=self.body,
                subject=self.subject
                )


        def make_files(self):
            """ Create and return Array of File objects of the Request. """
            # Check if attachment file has been defined as it is optional
            if self.attach_file != '':
                return [File(self.attach_file, 'file')]
            return []


    class Response(CloudResponse):
        """ Email Send Cloud Response object.

        EmailSend.Response
        """
        def __init__(self, **kwargs):
            """!
            Constructor

            @param **kwargs - Keyword arguments. Apply values to the request attributes.
                - @ref error
            """

            ## Error message
            self.error = ''
            super(EmailSend.Response, self).__init__(**kwargs)


    def __init__(self, **kwargs):
        """!
        Constructor

        @param **kwargs - Keyword argumen.ts. Apply values to the request attributes.
            - @ref Request.email
            - @ref Request.password
            - @ref Request.server
            - @ref Request.port
            - @ref Request.recipients
            - @ref Request.body
            - @ref Request.subject
            - @ref Request.attach_file
        """

        # Create and hold the Request object for this CloudMsg
        self.req = EmailSend.Request()
        # Create and hold the Response object for this CloudMsg
        self.resp = EmailSend.Response()
        super(EmailSend, self).__init__(
            svcname='email_send', **kwargs)

File: js/lib/cloud/email_send.js
#!/usr/bin/env node

var path = require('path');
var __cloudDir = path.join(__dirname);
var RAPPCloud = require(path.join(__cloudDir, 'RAPPCloud.js'));

/**
 * @fileOverview Prototype the RAPPCloud Service Method.
 * 
 * @class email_send
 * @memberof RAPPCloud
 * @description Asynchronous Service which will send an email to a list of recipients
 * @version 0.7.5
 * @author Lazaros Penteridis <lp@ortelio.co.uk>
 * @param email (String): The user's email username
 * @param passwd (String): The user's email password
 * @param server (String): The email server's smtp address, i.e. 'smtp.gmail.com'
 * @param port (String): The email server smtp port
 * @param recipients (Array): Email addresses of the recipients.
 * @param body (String): The email body
 * @param subject (String): The email subject
 * @param file(String): Path to a file attachment. Single file. In case of 
 * multiple attachments a zip file must be send to this field name.
 */
RAPPCloud.prototype.email_send = function(
                                            email,
                                            passwd,
                                            server,
                                            port,
                                            recipients,
                                            body,
                                            subject,
                                            file
                                         )
{
    var formData = require('form-data');
    var fs = require('fs');
    var cloud = this;
    var form = new formData();
    var request = cloud.determine_protocol();
    var body_obj = {};
    body_obj.email = cloud.escape_string(email);
    body_obj.passwd = cloud.escape_string(passwd);
    body_obj.server = cloud.escape_string(server);
    body_obj.port = cloud.escape_string(port);
    body_obj.recipients = [];
    for (var i=0; i<recipients.length; i++)
        body_obj.recipients[i] = cloud.escape_string(recipients[i]);
    body_obj.body = body;
    body_obj.subject = subject;
    var body_json = JSON.stringify(body_obj);
    
    if (!!file) {
       form.append('file', fs.createReadStream(file), { 
        filename: file
        }); 
    }
    
    form.append('json', body_json);
    var r = request.post(cloud.cloud_url + '/hop/email_send/ ', function(error, res, json){ 
        if (res.statusCode==200 && !error) {
            handle_reply(json);
        }
        else if (error) {
            error_handler(error);   
        }
        else if (res.statusCode != 200) {
            console.log(res.statusCode);
        }
    });
    r._form = form;
    r.setHeader('Connection', 'close');
    r.setHeader('Accept-Token', cloud.token);
    
    function handle_reply(json)
    {
		var json_obj;
		try {
			json_obj = JSON.parse(json);
			// JSON reply is: { error: '' }
		
			if(json_obj.error){  // Check for Errors  
				console.log('email_send JSON error: ' + json_obj.error);
			}
		} catch (e) {
			console.log('email_send::handle_reply Error parsing: ');
			return console.error(e);
		}
	}
	
	function error_handler(error) {
		return console.error(error);
	}   
};

/// Export
module.exports = RAPPCloud.email_send;
